### DISCUSSION TITLE
How to access, and set value of, @Environment(\.editMode) for a view in a test?

### DISCUSSION THREAD

**Discussion by grantneufeld** [nalexn/ViewInspector#257] [Q&A]
On iOS, I have a view accessing SwiftUI’s built-in `editMode` environment variable. I want to test based on different states (where `self.editMode.wrappedValue == .active`, and where `self.editMode.wrappedValue == .inactive`).

Assuming there is a way to properly access the editMode environment variable using ViewInspector, how can I change the value of that variable so I can test how my view responds to different states?

```swift
// MyView.swift
struct MyView: View {
    @Environment(\.editMode) private var editMode
    #if DEBUG
    internal var didAppear: ((Self) -> Void)?
    #endif
    var body: some View {
        Text(self.editMode != nil && self.editMode!.wrappedValue == .active ? "Active" : "Inactive")
        #if DEBUG
        .onAppear { self.didAppear?(self) }
        #endif
    }
}
```

```swift
// Test Attempt 1:
final class MyViewTests: XCTestCase {
    func test_editMode() throws {
        var sut = MyView()
        let expect = sut.on(\.didAppear) { view in // 🛑
            let inspectedEditMode = try view.environment(\.editMode)
            XCTAssertEqual(inspectedEditMode?.wrappedValue, .inactive)
        }
        ViewHosting.host(view: sut)
        wait(for: [expect], timeout: 0.1)
    }
}
```

Throws an error (at the line marked with 🛑): `failed - MyView does not have 'environment(Optional<Binding<EditMode>>)' modifier`

I then tried to provide the editMode:

```swift
// Test Attempt 2:
final class MyViewTests: XCTestCase {
    private var editMode: EditMode = .inactive
    private var editModeBinding: Binding<EditMode>?

    func test_editMode() throws {
        self.editMode = .active
        self.editModeBinding = Binding(wrappedValue: editMode)
        var sut = MyView()
            .environment(\.editMode, self.editModeBinding)
        // sut is no longer a `MyView`, now that it has the environment modifier
        // so can’t use the `didAppear` stuff
        let result = try sut.inspect().view(MyView.self).text().string()
        XCTAssertEqual(result, "Active") // 🛑 XCTAssertEqual failed: ("Inactive") is not equal to ("Active")
    }
}
```

Triggers a purple error in `MyView` where `body` is trying to access `self.editMode`:

🟪 `Accessing Environment<Optional<Binding<EditMode>>>'s value outside of being installed on a View. This will always read the default value and will not update.`

**Comment by nalexn** [comment]
Hey, I'm afraid `@Environment` properties aren't fully supported still: #131 ; this doesn't explain the error you see, but just letting you know even if you got it working, your end goal is currently unattainable

**Reply by grantneufeld** [reply]
Oh, that’s unfortunate. I hope Apple provides a way to access that for testing in the future!

Has there been a “Feedback Assistant” filed about this? (If so, what’s the FB#?)

**Reply by nalexn** [reply]
I don't think this would help at all. Apple's more likely to build a domestic testing framework for SwiftUI than adjust the inner workings of the framework for some third-party tools

**Comment by nalexn** [comment]
Closing. The main discussion is in #131
